source/extensions/filters/http/composite/filter.cc

```diff
@@ -12,6 +12,16 @@ namespace Composite {
 namespace {
 // Helper that returns `filter->func(args...)` if the filter is not null, returning `rval`
 // otherwise.
+template <class FilterPtrT, class FuncT, class RValT, class... Args>
+RValT delegateAllFilterActionOr(std::vector<FilterPtrT> filter_list, FuncT func, RValT rval, Args&&... args) {
+  RValT value = rval;
+  for (auto filter : filter_list) {
+    value = ((*filter).*func)(std::forward<Args>(args)...);
```

Comment:
How would this work if one of the intermediate filters rely on pausing filter iteration? Seems like we're going to return whatever the latest filter returns, which seems counter-intuitive

Reply:
I was testing locally and noticed this, would reverse order be preferable? I was thinking about possible alternatives to this, some sort of logic for chaining subsequent actions? Or alternatively, I think the best solution would maybe be to store the most recent status calls in the StreamFilterWrapper sub-filters and just return the last one.

Reply:
If you want this to work with any existing filter you'll end up having to maintain the existing behavior wrt filter control (pausing, resuming, etc), so I think you'll need to actually manage the filter start/stop/etc of each filter, possibly by reusing `FilterManager` to manage an internal set of filters

If you are ok with not supporting all filters this could be less powerful and we could document the limitations, but that makes this less useful than it could be

cc @alyssawilk @mattklein123 since filter iteration is complicated

Reply:
Coming in late to this but why are we reimplementing filter manager logic in a filter?  I'd lean towards reexamining the design here if that's the plan.

Reply:
@alyssawilk, we are attempting to implement a mechanism primarily to run multiple ext_proc filter calls per route, i.e. multiple actions per filter. I am relatively new to the envoy code base and was unaware of this in regards to the composite filter. 
That makes sense though.
I'll look into if this is possible through leveraging the `FilterManager` but another solution may be to come up with a different way to call multiple ext_proc filters.

Reply:
In general, it might be multiple filters and not just ext_proc, e.g. we have conditional execution of ext_authz and ext_proc interleaved in some scenarios we anticipate. 

Taking a step back, @snowp the original idea of composite filter was to enable essentially branching trees instead of a linear HTTP filter pipeline right? What we have here is something that looks more like:
```
-----> A --> B --> C --> [conditional execution of --> D --> E --> F] --> G
```

How would you approach this with composite filter? Should we have 3 composite filter instances and distribute the same conditional execution over each for [D, E, F]?

Reply:
Yeah the idea was to be able to go beyond a single linear set of filters, though at the time we only a use case for selecting one of N different filters based on the matching result. My original implementation for composite filters relied on https://github.com/envoyproxy/envoy/pull/14722, which would support multiple filters as the composite filter would inject new filters into the chain during matching. 

This was a big and scary change so we went with the simpler delegation approach instead since we had no use case at the time for a sequence of filters. With the delegate approach the composite filter proxies calls to the delegated filter, so there is no change to filter iteration state: the composite filter effectively becomes the delegated filter after matching. This doesn't naturally extend to multiple filters, since each filter in the chain expects to be able to modify the iteration state. 

Nowadays the FilterManager has been extracted out of the HCM and would possibly be used to manage the filter state, though it might end up being quite a bit of work. Inserting into the HCM filter manager is an option, but scarier as it involves modifying the HCM. 

So back to the original question I think stacking three composite filters next to each other is the easiest way to accomplish this right now. If the matching is expensive, we could make some low-risk changes to support propagating the match decision in FilterState/dynamic metadata to take the match result from the first filter and apply it to subsequent ones

Reply:
@alyssawilk  One thing I want to point out is that here we want to invoke the filter based on match result. For example,  based on matching with the request header. I don't think currently filter manager can do this.
  
Besides stacking three composite filters next to each other,  we probably can do it without composite filter. We can just leverage match_delegate: insert a list of ExtensionWithMatchers (each consist of a filter and matcher). Whether the filter will be invoked or not is based on matching condition.

Reply:
Oh I'd assumed the composite filter handled per-route matching that not ever filter had implemened. So if we waned
-----> A --> B --> C --> [conditional execution of --> D --> E --> F] --> G
do we have 
CompositeFilter [D E F] or
CompositeFilter[D] CompositeFilter[E]  CompositeFilter[F] 
I assume config is simpler and more elegant for the first, but then requires the composite filter to operate as a filter manager, handling all the pausing and resumption for D/E/F, vs the config complexity of the second coming with the simplicity of leveraging the HCM/upstream filter managers.  
If we can already do the latter with pure matching logic I really don't understand why we need the composite filter at all.  Having a filter chain in the filter chain seems to be more undesirable complexity and I think we could come up with cleaner API solutions for custom filter chains, like route based filter chain creation rather than having potentially infinitely layered filter managers

Reply:
Also to be clear I've put way less thought into this than other folks have, so please don't read my preference as an absolute - I'm very willing to be convinced, I'm quite possibly missing the benefits of the other options :-)

Reply:
I tend to agree with @alyssawilk here. At this point, we will align within Google on what our preferred option is and ping this PR when ready.

Reply:
Talked to Tianyu about this yesterday and I'm happy to whiteboard data plane options/complexity with anyone else interested.  As usual I'll defer to Harvey/Adi if we go the control plane route :-)

Reply:
The most interesting option that I don't fully understand is "dynamic HTTP filter chains", where somehow we augment the HTTP filter chain instantiation with per-route configuration. @alyssawilk do you have links or a dump of history on this one?

Reply:
alas I don't have context on that one.  I will say being able to enable and disable filters per route, or do custom filter chains per route seems entirely reasonable and if Envoy currently isn't flexible enough to do that in some dimension I'd be interested in making sure we fix that.  I just think there's easier ways than having layered filter managers :-)

Reply:
Here https://github.com/envoyproxy/envoy/pull/14722 and this doc https://docs.google.com/document/d/1dOa_v1eMA4F61f0jdFkxtbcO-_TYD76oMfGoY7nzZhk/edit#heading=h.9v2rpr8krgni might be able to provide some context.

Reply:
There has been asks for this for a while though I don't think we ever landed on a robust solution for it since there are a bunch of edge cases to consider, like what happens if the route changes mid-filter

Composite filters were an attempt at solving the "optional/branching filter" idea, with a focus on being deliverable via ECDS for a single filter. The idea was to let a separate system run an ECDS server and inject logic into the filter chain (at a pre-specified position in the filter chain) and deliver the matching requirement independent of other configuration, hence bundling the matching + filter config into something ECDS could deliver.

If we wanted the match component to live in the route configuration, I think we could get something working relatively easily by making use of https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/common/matcher/action/v3/skip_action.proto#envoy-v3-api-msg-extensions-filters-common-matcher-action-v3-skipfilter and having it read the decision from the per filter configuration or metadata from the route action. You could imagine configuring a sequence of ext_proc filters with something like 

`skip D unless route.metadata == "doTheThing"`
`skip E unless route.metadata == "doTheThing"`
`skip F unless route.metadata == "doTheThing"`